packages/reporting/src/request/utils/chrome-storage-map.js

```diff
@@ -0,0 +1,280 @@
+    // at the end of the burst. The soft limit will result in that,
+    // while the hard limit mitigates the risk that the script
+    // gets killed before the data gets persisted.
+    //
+    // Rule of thumbs:
+    // * The soft limit should be lower then the hard limit
+    // * The hard limit should not be set too high. Remember, it is
+    //   the protection against the browser unpredictably killing
+    //   the execution.
+    this.softFlushIntervalInMs = softFlushIntervalInMs;
+    this.hardFlushIntervalInMs = hardFlushIntervalInMs;
+    this._scheduledFlush = null;
+    this._lastFlush = Date.now();
+    this._dirty = false;
+
+    // Assumption: there should be enough time during startup to load
+    // the persisted map. Otherwise, the state will be inconsistent
+    // whenever the script is loaded (it will eventually become consistent,
+    // but that will not help if the browser kills it quickly).
+    //
+    // (If that assumption does not hold, _warnIfOutOfSync will detect
+    // and log it. A potential improvement could be to treat the
+    // in-memory map as the source of truth in that scenario.)
+    this.isReady = new Promise((resolve, reject) => {
+      chrome.storage.session.get([this.storageKey], (result) => {
+        if (chrome.runtime.lastError) {
+          reject(chrome.runtime.lastError);
+        } else {
+          const { entries = {}, ttl = {} } = result[this.storageKey] || {};
+          this._inMemoryMap = new Map(Object.entries(entries));
+          this._ttlMap = new Map(Object.entries(ttl));
+          this._initialSyncComplete = true;
+          this._expireOldEntries();
+          resolve();
+        }
+      });
+    });
+  }
+
+  get(_key) {
+    this._warnIfOutOfSync();
+    const key = this.normalizeKey(_key);
+    if (this._expireOldEntry(key)) {
+      return;
+    }
+    return this._inMemoryMap.get(key);
+  }
+
+  entries() {
+    this._warnIfOutOfSync();
+    this._expireOldEntries();
+    return this._inMemoryMap.entries();
+  }
+
+  values() {
+    this._warnIfOutOfSync();
+    this._expireOldEntries();
+    return Object.values(this._inMemoryMap);
+  }
+
+  has(_key) {
+    this._warnIfOutOfSync();
+    const key = this.normalizeKey(_key);
+    if (this._expireOldEntry(key)) {
+      return;
+    }
+    return this._inMemoryMap.has(key);
+  }
+
+  forEach(callback) {
+    this._warnIfOutOfSync();
+    this._expireOldEntries();
+    this._inMemoryMap.forEach(callback);
+  }
+
+  set(_key, value) {
+    this._warnIfOutOfSync();
+
+    // This should never trigger. Yet if the maps run full (perhaps
+    // as a side-effect of a bug), better reset then continuing with
+    // these huge maps.
+    if (
+      this._inMemoryMap.size >= this.maxEntries ||
+      this._ttlMap.size >= this.maxEntries
+    ) {
+      console.warn(
+        'AutoSyncingMap: Maps are running full (maybe you found a bug?). Purging data to prevent performance impacts.',
```

Comment:
Still uses the name "AutoSyncingMap", which conflicts with the existing data structure. Also, arguably "logger" could be better than console here.